node/impl/full/eth.go

```diff
@@ -1027,6 +1028,126 @@ func (a *EthModule) EthTraceTransaction(ctx context.Context, txHash string) ([]*
 	return txTraces, nil
 }
 
+func (a *EthModule) EthBlockNumberFromString(ctx context.Context, block string) (ethtypes.EthUint64, error) {
+
+	head := a.Chain.GetHeaviestTipSet()
+	switch block {
+	case "earliest":
+		return 0, xerrors.Errorf("block param \"earliest\" is not supported")
+	case "pending":
+		return ethtypes.EthUint64(head.Height()), nil
+	case "latest":
+		parent, err := a.Chain.GetTipSetFromKey(ctx, head.Parents())
+		if err != nil {
+			return 0, fmt.Errorf("cannot get parent tipset")
+		}
+		return ethtypes.EthUint64(parent.Height()), nil
+	case "safe":
+		latestHeight := head.Height() - 1
+		safeHeight := latestHeight - ethtypes.SafeEpochDelay
+		return ethtypes.EthUint64(safeHeight), nil
+	default:
+		blockNum, err := ethtypes.EthUint64FromHex(block)
+		if err != nil {
+			return 0, xerrors.Errorf("cannot parse fromBlock: %w", err)
+		}
+		return blockNum, err
+	}
+}
+
+func (a *EthModule) EthTraceFilter(ctx context.Context, filter ethtypes.EthTraceFilterCriteria) ([]*ethtypes.EthTraceFilterResult, error) {
+	fromBlock, err := a.EthBlockNumberFromString(ctx, filter.FromBlock)
+	if err != nil {
+		return nil, xerrors.Errorf("cannot parse fromBlock: %w", err)
+	}
+
+	toBlock, err := a.EthBlockNumberFromString(ctx, filter.ToBlock)
+	if err != nil {
+		return nil, xerrors.Errorf("cannot parse toBlock: %w", err)
+	}
+
+	var results []*ethtypes.EthTraceFilterResult
+	traceCounter := 0
+
+	for blkNum := fromBlock; blkNum <= toBlock; blkNum++ {
+		blockTraces, err := a.EthTraceBlock(ctx, strconv.FormatUint(uint64(blkNum), 10))
+		if err != nil {
+			return nil, xerrors.Errorf("cannot get trace for block %d: %w", blkNum, err)
+		}
+
+		for _, blockTrace := range blockTraces {
+			if matchFilterCriteria(blockTrace, filter) {
+				traceCounter++
+				if traceCounter <= filter.After {
```

Comment:
does this mean `After` is 1-based? so if I say "after 1" then I should expect the 2nd and beyond?

Reply:
see comment on limit offset